packages/dds/sequence/src/revertibles.ts

```diff
@@ -129,6 +153,110 @@ export function appendLocalPropertyChangedToRevertibles(
+			endIntervals.push({ offset: segmentLengths + interval.end.getOffset(), interval });
+			return true;
+		}
+	}
+	return false;
+}
+
+export function appendSharedStringDeltaToRevertibles(
+	string: SharedString,
+	delta: SequenceDeltaEvent,
+	revertibles: SharedStringRevertible[],
+) {
+	if (delta.ranges.length === 0) {
+		return;
+	}
+	if (delta.deltaOperation === MergeTreeDeltaType.REMOVE) {
+		const startIntervals: { offset: number; interval: SequenceInterval }[] = [];
+		const endIntervals: { offset: number; interval: SequenceInterval }[] = [];
+		let segmentLengths = 0;
+
+		// find interval endpoints in each segment
+		for (const deltaRange of delta.ranges) {
+			const refs = deltaRange.segment.localRefs;
+			if (refs !== undefined && deltaRange.position !== -1) {
+				for (const ref of refs) {
+					addIfIntervalEndpoint(ref, segmentLengths, startIntervals, endIntervals);
+				}
+			}
+			segmentLengths += deltaRange.segment.cachedLength;
+		}
+
+		if (startIntervals.length > 0 || endIntervals.length > 0) {
+			const removeRevertibles: MergeTreeDeltaRevertible[] = [];
+			appendToMergeTreeDeltaRevertibles(string, delta.deltaArgs, removeRevertibles);
+			assert(removeRevertibles.length === 1, "Remove revertible should be a single delta");
+
+			const revertible: TypedRevertible<typeof IntervalEventType.POSITIONREMOVE> = {
+				event: IntervalEventType.POSITIONREMOVE,
+				intervals: [],
+				mergeTreeRevertible: removeRevertibles[0],
+			};
+
+			// add an interval for each startInterval, accounting for any corresponding endIntervals
+			startIntervals.forEach(({ interval, offset }) => {
+				// find any corresponding end for this interval
+				const endIntervalIndex = endIntervals.findIndex((end) => {
+					return end.interval === interval;
+				});
+				let endOffset: number | undefined;
+				if (endIntervalIndex !== -1) {
+					endOffset = endIntervals[endIntervalIndex].offset;
+					endIntervals.splice(endIntervalIndex, 1);
+				}
+
+				revertible.intervals.push({
+					// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+					intervalId: interval.getIntervalId()!,
+					label: interval.properties.referenceRangeLabels[0],
+					startOffset: offset,
+					endOffset,
+				});
+			});
+
+			// add any remaining endIntervals that aren't matched with a startInterval
+			endIntervals.forEach(({ interval, offset }) => {
+				revertible.intervals.push({
+					// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+					intervalId: interval.getIntervalId()!,
+					label: interval.properties.referenceRangeLabels[0],
+					endOffset: offset,
+				});
+			});
+
+			revertibles.push(revertible);
+			return;
+		}
+	}
+
+	// Handle any merge tree delta that is not REMOVE or is REMOVE with no interval endpoints
+	const mergeTreeRevertibles: MergeTreeDeltaRevertible[] = [];
+	// Allow merging MergeTreeDeltaRevertible with previous
+	if (revertibles.length > 0 && "operation" in revertibles[revertibles.length - 1]) {
```

Comment:
maybe add a typeguard into the merge tree revertibles file, `isMergeTreeRevertible(maybe: unknonw): maybe is MergeTreeRevertible`, seem like it could be useful in general